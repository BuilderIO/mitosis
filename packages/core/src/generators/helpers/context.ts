import { MitosisComponent } from '../../types/mitosis-component';

export const hasContext = (component: MitosisComponent) =>
  Boolean(
    Object.keys(component.context.get).length ||
      Object.keys(component.context.set).length,
  );
