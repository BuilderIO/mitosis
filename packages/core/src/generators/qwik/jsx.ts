import { MitosisNode } from '../../types/mitosis-node';
import { DIRECTIVES } from './directives';
import {
  File,
  INDENT,
  invoke,
  NL,
  SrcBuilder,
  quote,
  UNINDENT,
  lastProperty,
} from './src-generator';
import { CssStyles } from './styles';

export function renderJSXNodes(
  file: File,
  directives: Map<string, string>,
  handlers: Map<string, string>,
  children: MitosisNode[],
  styles: Map<string, CssStyles>,
  parentSymbolBindings: Record<string, string>,
  root = true,
): any {
  return function(this: SrcBuilder) {
    if (children.length == 0) return;
    if (root) this.emit('(', INDENT, NL);
    const needsFragment = root && children.length > 1;
    file.import(file.qwikModule, 'h');
    if (needsFragment) {
      file.import(file.qwikModule, 'Fragment');
      this.jsxBeginFragment(file.import(file.qwikModule, 'Fragment'));
    }
    children.forEach((child) => {
      if (isEmptyTextNode(child)) return;
      if (isTextNode(child)) {
        if (child.bindings._text !== undefined) {
          this.jsxTextBinding(child.bindings._text);
        } else {
          this.isJSX
            ? this.emit(child.properties._text)
            : this.jsxTextBinding(quote(child.properties._text!));
        }
      } else {
        let childName = child.name;
        const directive = DIRECTIVES[childName];
        if (typeof directive == 'function') {
          this.emit(
            directive(child, () =>
              renderJSXNodes(
                file,
                directives,
                handlers,
                child.children,
                styles,
                {},
                false,
              ).call(this),
            ),
          );
        } else {
          if (typeof directive == 'string') {
            directives.set(childName, directive);
            if (file.module !== 'med') {
              file.import('./med.js', childName);
            }
          }
          if (isSymbol(childName)) {
            // TODO(misko): We are hard coding './med.js' which is not right.
            file.import('./med.js', childName);
            let exportedChildName = file.exports.get(childName);
            if (exportedChildName) {
              childName = exportedChildName;
            }
          }
          let props: Record<string, any> = child.properties;
          const css = child.bindings.css;
          if (css) {
            props = { ...props };
            props.class = addClass(styles.get(css)!.CLASS_NAME, props.class);
          }
          const symbolBindings: Record<string, string> = {};
          const bindings = rewriteHandlers(
            file,
            handlers,
            child.bindings,
            symbolBindings,
          );
          this.jsxBegin(childName, props, {
            ...bindings,
            ...parentSymbolBindings,
          });
          renderJSXNodes(
            file,
            directives,
            handlers,
            child.children,
            styles,
            symbolBindings,
            false,
          ).call(this);
          this.jsxEnd(childName);
        }
      }
    });
    if (needsFragment) {
      this.jsxEndFragment();
    }
    if (root) this.emit(UNINDENT, ')');
  };
}

function isSymbol(name: string): boolean {
  return name.charAt(0) == name.charAt(0).toUpperCase();
}

function addClass(
  className: string,
  existingClass: string | undefined,
): string {
  return [className, ...(existingClass ? existingClass.split(' ') : [])].join(
    ' ',
  );
}

function isEmptyTextNode(child: MitosisNode) {
  return child.properties._text?.trim() == '';
}

function isTextNode(child: MitosisNode) {
  return (
    child.properties._text !== undefined || child.bindings._text !== undefined
  );
}

/**
 * Rewrites bindings:
 * - Remove `css`
 * - Rewrites event handles
 * - Extracts symbol bindings.
 *
 * @param file
 * @param handlers
 * @param bindings
 * @param symbolBindings Options record which will receive the symbol bindings
 * @returns
 */
function rewriteHandlers(
  file: File,
  handlers: Map<string, string>,
  bindings: Record<string, string | undefined>,
  symbolBindings: Record<string, string>,
): Record<string, string> {
  const outBindings: Record<string, string> = {};
  for (let key in bindings) {
    if (Object.prototype.hasOwnProperty.call(bindings, key)) {
      let binding: any = bindings[key]!;
      let handlerBlock: string | undefined;
      if (binding != null) {
        if (key == 'css') {
          continue;
        } else if ((handlerBlock = handlers.get(binding))) {
          key = `on:${key.substring(2).toLowerCase()}`;
          binding = invoke(file.import(file.qwikModule, 'qrl'), [
            quote(file.qrlPrefix + 'high.js'),
            quote(handlerBlock),
            '[__props__, __state__]',
          ]);
        } else if (symbolBindings && key.startsWith('symbol.data.')) {
          symbolBindings[lastProperty(key)] = binding;
        }
        outBindings[key] = binding;
      }
    }
  }
  return outBindings;
}
