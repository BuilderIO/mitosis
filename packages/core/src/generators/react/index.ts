export * from './generator';
export * from './types';
