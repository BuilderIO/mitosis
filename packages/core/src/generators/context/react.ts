import { format } from 'prettier/standalone';
import { getMemberObjectString } from '../../helpers/get-state-object-string';
import { MitosisContext } from '../../types/mitosis-context';

type ContextToReactOptions = {
  format?: boolean;
};

export const contextToReact =
  (options: ContextToReactOptions = {}) =>
  ({ context }: { context: MitosisContext }): string => {
    let str = `
  import { createContext } from 'react';

  export default createContext(${getMemberObjectString(context.value)})
  `;

    if (options.format !== false) {
      try {
        str = format(str, {
          parser: 'typescript',
          plugins: [
            require('prettier/parser-typescript'), // To support running in browsers
          ],
        });
      } catch (err) {
        console.error('Format error for file:', str);
        throw err;
      }
    }

    return str;
  };
