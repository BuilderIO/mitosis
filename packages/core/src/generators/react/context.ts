import { contextPropDrillingKey, MitosisComponent, ToReactOptions } from '@builder.io/mitosis';
import { stringifyContextValue } from '../../helpers/get-state-object-string';
import { createMitosisNode } from '../../helpers/create-mitosis-node';
import { createSingleBinding } from '../../helpers/bindings';

export function provideContext(
  json: MitosisComponent,
  options: Pick<ToReactOptions, 'contextType'>,
): string | void {
  if (options.contextType === 'prop-drill') {
    let str = '';
    for (const key in json.context.set) {
      const { name, ref, value } = json.context.set[key];
      if (value) {
        str += `
          ${contextPropDrillingKey}.${name} = ${stringifyContextValue(value)};
        `;
      }
      // TODO: support refs. I'm not sure what those are so unclear how to support them
    }
    return str;
  } else {
    for (const key in json.context.set) {
      const { name, ref, value } = json.context.set[key];
      if (value) {
        json.children = [
          createMitosisNode({
            name: `${name}.Provider`,
            children: json.children,
            ...(value && {
              bindings: {
                value: createSingleBinding({
                  code: stringifyContextValue(value),
                }),
              },
            }),
          }),
        ];
      } else if (ref) {
        json.children = [
          createMitosisNode({
            name: 'Context.Provider',
            children: json.children,
            ...(ref && {
              bindings: {
                value: createSingleBinding({ code: ref }),
              },
            }),
          }),
        ];
      }
    }
  }
}
