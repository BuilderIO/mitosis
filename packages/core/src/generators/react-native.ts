import json5 from 'json5';
import { camelCase, size } from 'lodash';
import { fastClone } from '../helpers/fast-clone';
import traverse from 'traverse';
import { ClassStyleMap } from '../helpers/collect-styles';
import { isMitosisNode } from '../helpers/is-mitosis-node';
import { MitosisComponent } from '../types/mitosis-component';
import { componentToReact } from './react';
import { BaseTranspilerOptions, Transpiler } from '../types/config';

export interface ToReactNativeOptions extends BaseTranspilerOptions {
  stylesType?: 'emotion' | 'react-native';
  stateType?: 'useState' | 'mobx' | 'valtio' | 'solid' | 'builder';
}

const stylePropertiesThatMustBeNumber = new Set(['lineHeight']);

export const collectReactNativeStyles = (
  json: MitosisComponent,
): ClassStyleMap => {
  const styleMap: ClassStyleMap = {};

  const componentIndexes: { [className: string]: number | undefined } = {};

  traverse(json).forEach(function(item) {
    if (isMitosisNode(item)) {
      if (typeof item.bindings.css === 'string') {
        const value = json5.parse(item.bindings.css);
        delete item.bindings.css;
        if (!size(value)) {
          return;
        }

        // Style properties like `"20px"` need to be numbers like `20` for react native
        for (const key in value) {
          const propertyValue = value[key];
          if (typeof propertyValue === 'string' && propertyValue.match(/^\d/)) {
            const newValue = parseFloat(propertyValue);
            if (!isNaN(newValue)) {
              value[key] = newValue;
            }
          }
          if (
            stylePropertiesThatMustBeNumber.has(key) &&
            typeof value[key] !== 'number'
          ) {
            console.warn(
              `Style key ${key} must be a number, but had value \`${value[key]}\``,
            );
            delete value[key];
          }
        }
        const componentName = camelCase(item.name || 'view');
        const index = (componentIndexes[componentName] =
          (componentIndexes[componentName] || 0) + 1);
        const className = `${componentName}${index}`;
        item.bindings.style = `styles.${className}`;

        styleMap[className] = value;
      }
    }
  });

  return styleMap;
};

export const componentToReactNative = (
  options: ToReactNativeOptions = {},
): Transpiler => ({ component, path }) => {
  const json = fastClone(component);
  traverse(json).forEach((node) => {
    if (isMitosisNode(node)) {
      // TODO: handle TextInput, Image, etc
      if (node.name.toLowerCase() === node.name) {
        node.name = 'View';
      }

      if (
        node.properties._text?.trim().length ||
        node.bindings._text?.trim()?.length
      ) {
        node.name = 'Text';
      }
    }
  });

  return componentToReact({
    ...options,
    stylesType: options.stylesType || 'react-native',
    type: 'native',
  })({ component: json, path });
};
