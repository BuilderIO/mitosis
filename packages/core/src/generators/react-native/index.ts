import json5 from 'json5';
import { camelCase, size } from 'lodash';
import { fastClone } from '../../helpers/fast-clone';
import traverse from 'traverse';
import { ClassStyleMap } from '../../helpers/styles/helpers';
import { isMitosisNode } from '../../helpers/is-mitosis-node';
import { MitosisComponent } from '../../types/mitosis-component';
import { componentToReact } from '../react';
import { BaseTranspilerOptions, TranspilerGenerator } from '../../types/transpiler';
import { MitosisNode, Plugin } from '../..';
import { createSingleBinding } from '../../helpers/bindings';
import { Dictionary } from '../../helpers/typescript';
import { mergeOptions } from '../../helpers/merge-options';
import isChildren from '../../helpers/is-children';
import { sanitizeReactNativeBlockStyles } from './sanitize-react-native-block-styles';

export interface ToReactNativeOptions extends BaseTranspilerOptions {
  stylesType: 'emotion' | 'react-native';
  stateType: 'useState' | 'mobx' | 'valtio' | 'solid' | 'builder';
}

const stylePropertiesThatMustBeNumber = new Set(['lineHeight']);

const MEDIA_QUERY_KEY_REGEX = /^@media.*/;

const sanitizeStyle = (obj: any) => (key: string, value: string) => {
  const propertyValue = obj[key];

  if (key.match(MEDIA_QUERY_KEY_REGEX)) {
    console.warn('Unsupported: skipping media queries for react-native: ', key, propertyValue);
    delete obj[key];
    return;
  }
};

export const collectReactNativeStyles = (json: MitosisComponent): ClassStyleMap => {
  const styleMap: ClassStyleMap = {};

  const componentIndexes: Dictionary<number | undefined> = {};
  const getStyleSheetName = (item: MitosisNode) => {
    const componentName = camelCase(item.name || 'view');
    // If we have already seen this component name, we will increment the index. Otherwise, we will set the index to 1.
    const index = (componentIndexes[componentName] = (componentIndexes[componentName] || 0) + 1);
    return `${componentName}${index}`;
  };
  traverse(json).forEach(function (item) {
    if (!isMitosisNode(item)) {
      return;
    }
    let cssValue = json5.parse(item.bindings.css?.code || '{}');
    delete item.bindings.css;

    if (size(cssValue)) {
      // Style properties like `"20px"` need to be numbers like `20` for react native
      for (const key in cssValue) {
        sanitizeStyle(cssValue)(key, cssValue[key]);
        cssValue = sanitizeReactNativeBlockStyles(cssValue);
      }
    }

    try {
      let styleValue = json5.parse(item.bindings.style?.code || '{}');
      if (size(styleValue)) {
        // Style properties like `"20px"` need to be numbers like `20` for react native
        for (const key in styleValue) {
          sanitizeStyle(styleValue)(key, styleValue[key]);
          styleValue = sanitizeReactNativeBlockStyles(styleValue);
        }

        item.bindings.style!.code = json5.stringify(styleValue);
      }
    } catch (e) {}

    if (!size(cssValue)) {
      return;
    }

    const styleSheetName = getStyleSheetName(item);
    const styleSheetAccess = `styles.${styleSheetName}`;

    item.bindings.style = createSingleBinding({
      code:
        item.bindings.style?.code.replace(/}$/, `, ...${styleSheetAccess} }`) || styleSheetAccess,
    });

    styleMap[styleSheetName] = cssValue;
  });

  return styleMap;
};

/**
 * Plugin that handles necessary transformations from React to React Native:
 * - Converts DOM tags to <View /> and <Text />
 * - Removes redundant `class`/`className` attributes
 */
const PROCESS_REACT_NATIVE_PLUGIN: Plugin = () => ({
  json: {
    pre: (json: MitosisComponent) => {
      traverse(json).forEach((node) => {
        if (isMitosisNode(node)) {
          // TODO: handle TextInput, Image, etc
          if (isChildren({ node })) {
            node.name = '';
          } else if (node.name.toLowerCase() === node.name) {
            node.name = 'View';
          } else if (
            node.properties._text?.trim().length ||
            node.bindings._text?.code?.trim()?.length
          ) {
            node.name = 'Text';
          }

          if (node.properties.class) {
            delete node.properties.class;
          }
          if (node.properties.className) {
            delete node.properties.className;
          }
          if (node.bindings.class) {
            delete node.bindings.class;
          }
          if (node.bindings.className) {
            delete node.bindings.className;
          }
        }
      });
    },
  },
});

const DEFAULT_OPTIONS: ToReactNativeOptions = {
  stateType: 'useState',
  stylesType: 'react-native',
  plugins: [PROCESS_REACT_NATIVE_PLUGIN],
};

export const componentToReactNative: TranspilerGenerator<Partial<ToReactNativeOptions>> =
  (_options = {}) =>
  ({ component, path }) => {
    const json = fastClone(component);

    const options = mergeOptions(DEFAULT_OPTIONS, _options);

    return componentToReact({ ...options, type: 'native' })({ component: json, path });
  };
