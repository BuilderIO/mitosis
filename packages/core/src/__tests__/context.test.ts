import { parseJsx } from '../parsers/jsx';
import { contextToReact } from '../generators/context/react';
import { parseContext } from '../parsers/context';
import { componentToReact } from '../generators/react';
import { componentToReactNative } from '../generators/react-native';

const simpleExample = require('./data/context/simple.context.lite');
const componentWithContext = require('./data/context/component-with-context.lite');
const renderBlock = require('./data/blocks/builder-render-block.raw');

describe('Context', () => {
  test('Parse context', () => {
    const component = parseContext(simpleExample, { name: 'SimpleExample' });
    if (!component) {
      throw new Error('No parseable context found for simple.context.lite.ts');
    }
    expect(component).toMatchSnapshot();
    const reactContext = contextToReact()({ context: component });
    expect(reactContext).toMatchSnapshot();
  });

  test('Use and set context in components', () => {
    const component = parseJsx(componentWithContext);
    expect(component).toMatchSnapshot();
    const reactComponent = componentToReact()({ component });
    expect(reactComponent).toMatchSnapshot();

    const reactNativeComponent = componentToReactNative()({ component });
    expect(reactNativeComponent).toMatchSnapshot();
  });

  test('Use and set context in complex components', () => {
    const component = parseJsx(renderBlock);
    expect(component).toMatchSnapshot();
    const reactComponent = componentToReact()({ component });
    expect(reactComponent).toMatchSnapshot();
  });
});
