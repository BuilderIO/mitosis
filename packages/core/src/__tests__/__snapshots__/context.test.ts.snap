// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context Parse context 1`] = `
Object {
  "@type": "@builder.io/mitosis/context",
  "name": "SimpleExample",
  "value": Object {
    "content": null,
    "context": Object {},
    "foo": "bar",
    "fooUpperCase": "@builder.io/mitosis/method:get fooUpperCase() {
  return this.foo.toUpperCase();
}",
    "someMethod": "@builder.io/mitosis/method:someMethod() {
  return this.fooUpperCase.toLowercase();
}",
    "state": Object {},
  },
}
`;

exports[`Context Parse context 2`] = `
"import { createContext } from \\"react\\";

export default createContext({
  foo: \\"bar\\",
  get fooUpperCase() {
    return this.foo.toUpperCase();
  },
  someMethod() {
    return this.fooUpperCase.toLowercase();
  },
  content: null,
  context: {},
  state: {},
});
"
`;

exports[`Context Use and set context in complex components 1`] = `
Object {
  "@type": "@builder.io/mitosis/component",
  "children": Array [
    Object {
      "@type": "@builder.io/mitosis/node",
      "bindings": Object {},
      "children": Array [
        Object {
          "@type": "@builder.io/mitosis/node",
          "bindings": Object {},
          "children": Array [],
          "meta": Object {},
          "name": "div",
          "properties": Object {
            "_text": "
      ",
          },
        },
        Object {
          "@type": "@builder.io/mitosis/node",
          "bindings": Object {},
          "children": Array [],
          "meta": Object {},
          "name": "div",
          "properties": Object {
            "_text": "
      ",
          },
        },
        Object {
          "@type": "@builder.io/mitosis/node",
          "bindings": Object {},
          "children": Array [],
          "meta": Object {},
          "name": "div",
          "properties": Object {
            "_text": "
      ",
          },
        },
        Object {
          "@type": "@builder.io/mitosis/node",
          "bindings": Object {
            "_spread": "state.properties",
            "style": "state.css",
          },
          "children": Array [
            Object {
              "@type": "@builder.io/mitosis/node",
              "bindings": Object {},
              "children": Array [],
              "meta": Object {},
              "name": "div",
              "properties": Object {
                "_text": "
        ",
              },
            },
            Object {
              "@type": "@builder.io/mitosis/node",
              "bindings": Object {
                "block": "state.useBlock",
              },
              "children": Array [],
              "meta": Object {},
              "name": "BlockStyles",
              "properties": Object {},
            },
            Object {
              "@type": "@builder.io/mitosis/node",
              "bindings": Object {},
              "children": Array [],
              "meta": Object {},
              "name": "div",
              "properties": Object {
                "_text": "
        ",
              },
            },
            Object {
              "@type": "@builder.io/mitosis/node",
              "bindings": Object {
                "when": "state.componentRef",
              },
              "children": Array [
                Object {
                  "@type": "@builder.io/mitosis/node",
                  "bindings": Object {
                    "_spread": "state.componentOptions",
                    "children": "state.useBlock.children",
                  },
                  "children": Array [],
                  "meta": Object {},
                  "name": "state.componentRef",
                  "properties": Object {},
                },
              ],
              "meta": Object {},
              "name": "Show",
              "properties": Object {},
            },
            Object {
              "@type": "@builder.io/mitosis/node",
              "bindings": Object {},
              "children": Array [],
              "meta": Object {},
              "name": "div",
              "properties": Object {
                "_text": "
        ",
              },
            },
            Object {
              "@type": "@builder.io/mitosis/node",
              "bindings": Object {
                "when": "!state.componentRef && state.useBlock.children && state.useBlock.children.length",
              },
              "children": Array [
                Object {
                  "@type": "@builder.io/mitosis/node",
                  "bindings": Object {},
                  "children": Array [],
                  "meta": Object {},
                  "name": "div",
                  "properties": Object {
                    "_text": "
          ",
                  },
                },
                Object {
                  "@type": "@builder.io/mitosis/node",
                  "bindings": Object {
                    "each": "state.useBlock.children",
                  },
                  "children": Array [
                    Object {
                      "@type": "@builder.io/mitosis/node",
                      "bindings": Object {
                        "block": "child",
                        "index": "index",
                      },
                      "children": Array [],
                      "meta": Object {},
                      "name": "RenderBlock",
                      "properties": Object {},
                    },
                  ],
                  "meta": Object {},
                  "name": "For",
                  "properties": Object {
                    "_forName": "child",
                  },
                },
                Object {
                  "@type": "@builder.io/mitosis/node",
                  "bindings": Object {},
                  "children": Array [],
                  "meta": Object {},
                  "name": "div",
                  "properties": Object {
                    "_text": "
        ",
                  },
                },
              ],
              "meta": Object {},
              "name": "Show",
              "properties": Object {},
            },
            Object {
              "@type": "@builder.io/mitosis/node",
              "bindings": Object {},
              "children": Array [],
              "meta": Object {},
              "name": "div",
              "properties": Object {
                "_text": "
      ",
              },
            },
          ],
          "meta": Object {},
          "name": "state.tagName",
          "properties": Object {},
        },
        Object {
          "@type": "@builder.io/mitosis/node",
          "bindings": Object {},
          "children": Array [],
          "meta": Object {},
          "name": "div",
          "properties": Object {
            "_text": "
      ",
          },
        },
        Object {
          "@type": "@builder.io/mitosis/node",
          "bindings": Object {},
          "children": Array [],
          "meta": Object {},
          "name": "div",
          "properties": Object {
            "_text": "
    ",
          },
        },
      ],
      "meta": Object {},
      "name": "Fragment",
      "properties": Object {},
    },
  ],
  "context": Object {
    "get": Object {
      "builderContext": Object {
        "name": "BuilderContext",
        "path": "../context/builder.context.lite:default",
      },
    },
    "set": Object {},
  },
  "hooks": Object {},
  "imports": Array [
    Object {
      "imports": Object {
        "getBlockComponentOptions": "getBlockComponentOptions",
      },
      "path": "../functions/get-block-component-options",
    },
    Object {
      "imports": Object {
        "getBlockProperties": "getBlockProperties",
      },
      "path": "../functions/get-block-properties",
    },
    Object {
      "imports": Object {
        "getBlockStyles": "getBlockStyles",
      },
      "path": "../functions/get-block-styles",
    },
    Object {
      "imports": Object {
        "getBlockTag": "getBlockTag",
      },
      "path": "../functions/get-block-tag",
    },
    Object {
      "imports": Object {
        "components": "components",
      },
      "path": "../functions/register-component",
    },
    Object {
      "imports": Object {
        "BuilderContext": "default",
      },
      "path": "../context/builder.context.lite",
    },
    Object {
      "imports": Object {
        "getBlockActions": "getBlockActions",
      },
      "path": "../functions/get-block-actions",
    },
    Object {
      "imports": Object {
        "getProcessedBlock": "getProcessedBlock",
      },
      "path": "../functions/get-processed-block",
    },
    Object {
      "imports": Object {
        "BlockStyles": "default",
      },
      "path": "./block-styles.lite",
    },
  ],
  "inputs": Array [],
  "meta": Object {},
  "name": "RenderBlock",
  "state": Object {
    "actions": "@builder.io/mitosis/method:get actions() {
  return getBlockActions({
    block: state.useBlock,
    state: builderContext.state,
    context: builderContext.context
  });
}",
    "component": "@builder.io/mitosis/method:get component() {
  const componentName = state.useBlock.component?.name;

  if (!componentName) {
    return null;
  }

  const ref = components[state.useBlock.component?.name!];

  if (componentName && !ref) {
    // TODO: Public doc page with more info about this message
    console.warn(\`
          Could not find a registered component named \\"\${componentName}\\".
          If you registered it, is the file that registered it imported by the file that needs to render it?\`);
  }

  return ref;
}",
    "componentInfo": "@builder.io/mitosis/method:get componentInfo() {
  return state.component?.info;
}",
    "componentOptions": "@builder.io/mitosis/method:get componentOptions() {
  return getBlockComponentOptions(state.useBlock);
}",
    "componentRef": "@builder.io/mitosis/method:get componentRef() {
  return state.component?.component;
}",
    "css": "@builder.io/mitosis/method:get css() {
  return getBlockStyles(state.useBlock);
}",
    "properties": "@builder.io/mitosis/method:get properties() {
  return getBlockProperties(state.useBlock);
}",
    "tagName": "@builder.io/mitosis/method:get tagName() {
  return (getBlockTag(state.useBlock) as any);
}",
    "useBlock": "@builder.io/mitosis/method:get useBlock() {
  return getProcessedBlock({
    block: props.block,
    state: builderContext.state,
    context: builderContext.context
  });
}",
  },
  "subComponents": Array [],
}
`;

exports[`Context Use and set context in complex components 2`] = `
"import { useContext } from \\"react\\";
import { getBlockComponentOptions } from \\"../functions/get-block-component-options\\";
import { getBlockProperties } from \\"../functions/get-block-properties\\";
import { getBlockStyles } from \\"../functions/get-block-styles\\";
import { getBlockTag } from \\"../functions/get-block-tag\\";
import { components } from \\"../functions/register-component\\";
import BuilderContext from \\"../context/builder.context\\";
import { getBlockActions } from \\"../functions/get-block-actions\\";
import { getProcessedBlock } from \\"../functions/get-processed-block\\";
import BlockStyles from \\"./block-styles.lite\\";

export default function RenderBlock(props) {
  function component() {
    const componentName = useBlock().component?.name;

    if (!componentName) {
      return null;
    }

    const ref = components[useBlock().component?.name];

    if (componentName && !ref) {
      // TODO: Public doc page with more info about this message
      console.warn(\`
        Could not find a registered component named \\"\${componentName}\\".
        If you registered it, is the file that registered it imported by the file that needs to render it?\`);
    }

    return ref;
  }

  function componentInfo() {
    return component?.()?.info;
  }

  function componentRef() {
    return component?.()?.component;
  }

  function tagName() {
    return getBlockTag(useBlock());
  }

  function properties() {
    return getBlockProperties(useBlock());
  }

  function useBlock() {
    return getProcessedBlock({
      block: props.block,
      state: builderContext.state,
      context: builderContext.context,
    });
  }

  function actions() {
    return getBlockActions({
      block: useBlock(),
      state: builderContext.state,
      context: builderContext.context,
    });
  }

  function css() {
    return getBlockStyles(useBlock());
  }

  function componentOptions() {
    return getBlockComponentOptions(useBlock());
  }

  const builderContext = useContext(BuilderContext);

  const TagNameRef = tagName();
  const ComponentRefRef = componentRef();

  return (
    <>
      <TagNameRef {...properties()} style={css()}>
        <BlockStyles block={useBlock()} />

        {componentRef() ? (
          <ComponentRefRef
            {...componentOptions()}
            children={useBlock().children}
          />
        ) : null}

        {!componentRef() &&
        useBlock().children &&
        useBlock().children.length ? (
          <>
            {useBlock().children?.map((child, index) => (
              <RenderBlock index={index} block={child} />
            ))}
          </>
        ) : null}
      </TagNameRef>
    </>
  );
}
"
`;

exports[`Context Use and set context in components 1`] = `
Object {
  "@type": "@builder.io/mitosis/component",
  "children": Array [
    Object {
      "@type": "@builder.io/mitosis/node",
      "bindings": Object {},
      "children": Array [
        Object {
          "@type": "@builder.io/mitosis/node",
          "bindings": Object {},
          "children": Array [],
          "meta": Object {},
          "name": "div",
          "properties": Object {
            "_text": "
      ",
          },
        },
        Object {
          "@type": "@builder.io/mitosis/node",
          "bindings": Object {},
          "children": Array [
            Object {
              "@type": "@builder.io/mitosis/node",
              "bindings": Object {
                "_text": "foo.value",
              },
              "children": Array [],
              "meta": Object {},
              "name": "div",
              "properties": Object {},
            },
          ],
          "meta": Object {},
          "name": "Fragment",
          "properties": Object {},
        },
        Object {
          "@type": "@builder.io/mitosis/node",
          "bindings": Object {},
          "children": Array [],
          "meta": Object {},
          "name": "div",
          "properties": Object {
            "_text": "
    ",
          },
        },
      ],
      "meta": Object {},
      "name": "Fragment",
      "properties": Object {},
    },
  ],
  "context": Object {
    "get": Object {
      "foo": Object {
        "name": "Context1",
        "path": "@dummy/1:default",
      },
    },
    "set": Object {
      "@dummy/1:default": Object {
        "name": "Context1",
        "value": Object {
          "foo": "bar",
        },
      },
      "@dummy/2:default": Object {
        "name": "Context2",
        "value": Object {
          "bar": "baz",
        },
      },
    },
  },
  "hooks": Object {},
  "imports": Array [
    Object {
      "imports": Object {
        "Context1": "default",
      },
      "path": "@dummy/1",
    },
    Object {
      "imports": Object {
        "Context2": "default",
      },
      "path": "@dummy/2",
    },
  ],
  "inputs": Array [],
  "meta": Object {},
  "name": "ComponentWithContext",
  "state": Object {},
  "subComponents": Array [],
}
`;

exports[`Context Use and set context in components 2`] = `
"import { useContext } from \\"react\\";
import Context1 from \\"@dummy/1\\";
import Context2 from \\"@dummy/2\\";

export default function ComponentWithContext(props) {
  const foo = useContext(Context1);

  return (
    <Context2.Provider
      value={{
        bar: \\"baz\\",
      }}
    >
      <Context1.Provider
        value={{
          foo: \\"bar\\",
        }}
      >
        <>{foo.value}</>
      </Context1.Provider>
    </Context2.Provider>
  );
}
"
`;

exports[`Context Use and set context in components 3`] = `
"import * as React from \\"react\\";
import { View, StyleSheet, Image, Text } from \\"react-native\\";
import { useContext } from \\"react\\";
import Context1 from \\"@dummy/1\\";
import Context2 from \\"@dummy/2\\";

export default function ComponentWithContext(props) {
  const foo = useContext(Context1);

  return (
    <Context2.Provider
      value={{
        bar: \\"baz\\",
      }}
    >
      <Context1.Provider
        value={{
          foo: \\"bar\\",
        }}
      >
        <>
          <Text>{foo.value}</Text>
        </>
      </Context1.Provider>
    </Context2.Provider>
  );
}
"
`;
